### DISCUSSION TITLE
Streamline position property of InputEvent classes

### DISCUSSION THREAD

**Discussion by Sauermann** [godotengine/godot-proposals#4158] [Engine Core]
Currently the `position` property of input events is defined in four different classes:
- InputEventScreenDrag
- InputEventScreenTouch
- InputEventGesture
- InputEventMouse

This can be streamlined, so that position is defined only in a single class and here is a proposal of how to do this:

- InputEventFromWindow
  - InputEventWithModifiers
    - InputEventKey
    - InputEventPosition (new)
      - InputEventScreenDrag
      - InputEventScreenTouch
      - InputEventGesture
      - InputEventMouse

Besides the intended change, this would add Modifiers (shift, ctrl, ...) to `InputEventScreenDrag` and `InputEventScreenTouch`. This would extend the functionality while simplifying the position-code.

**Comment by aaronfranke** [comment]
Do you have a practical use case in mind for this change? Is there any use case for doing `if input_event is InputEventPosition:` for example?

**Reply by Sauermann** [reply]
@aaronfranke In order to handle InputEvent processing in `SubViewportContainer` correctly, it is necessary to distinguish between events without and events with positions. My arguments for this distinction can be seen in the "Events without position parameter" section of my comment https://github.com/godotengine/godot/pull/55300#issuecomment-1065854555.

This proposal is a functionality, that would improve my PR https://github.com/godotengine/godot/pull/58334

The current best way I found to do this is
``` C++
if (p_event->has_method("set_position")){ ...
```
but this would depend on the function being actually exposed.

A different way, would be to check all four different classes, but this would be error prone, as soon as additional events with positions would be added.
